core/src/main/java/org/apache/iceberg/rest/RESTTableOperations.java

```diff
@@ -0,0 +1,212 @@
+ * under the License.
+ */
+
+package org.apache.iceberg.rest;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import org.apache.iceberg.LocationProviders;
+import org.apache.iceberg.MetadataUpdate;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.rest.requests.UpdateTableRequest;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.rest.responses.LoadTableResponse;
+
+class RESTTableOperations implements TableOperations {
+  private static final String METADATA_FOLDER_NAME = "metadata";
+
+  enum UpdateType {
+    CREATE,
+    REPLACE,
+    SIMPLE
+  }
+
+  private final RESTClient client;
+  private final String path;
+  private final FileIO io;
+  private final List<MetadataUpdate> createChanges;
+  private final TableMetadata replaceBase;
+  private UpdateType updateType;
+  private TableMetadata current;
+
+  RESTTableOperations(RESTClient client, String path, FileIO io, TableMetadata current) {
+    this(client, path, io, UpdateType.SIMPLE, Lists.newArrayList(), current);
+  }
+
+  RESTTableOperations(RESTClient client, String path, FileIO io, UpdateType updateType,
+                      List<MetadataUpdate> createChanges, TableMetadata current) {
+    this.client = client;
+    this.path = path;
+    this.io = io;
+    this.updateType = updateType;
+    this.createChanges = createChanges;
+    this.replaceBase = current;
+    if (updateType == UpdateType.CREATE) {
+      this.current = null;
+    } else {
+      this.current = current;
+    }
+  }
+
+  @Override
+  public TableMetadata current() {
+    return current;
+  }
+
+  @Override
+  public TableMetadata refresh() {
+    return updateCurrentMetadata(client.get(path, LoadTableResponse.class, ErrorHandlers.tableErrorHandler()));
+  }
+
+  @Override
+  public void commit(TableMetadata base, TableMetadata metadata) {
+    UpdateTableRequest.Builder requestBuilder;
+    List<MetadataUpdate> baseChanges;
+    Consumer<ErrorResponse> errorHandler;
+    switch (updateType) {
+      case CREATE:
+        Preconditions.checkState(base == null, "Invalid base metadata for create transaction, expected null: %s", base);
+        requestBuilder = UpdateTableRequest.builderForCreate();
+        baseChanges = createChanges;
+        errorHandler = ErrorHandlers.tableErrorHandler(); // throws NoSuchTableException
+        break;
+
+      case REPLACE:
+        Preconditions.checkState(base != null, "Invalid base metadata: null");
```

Comment:
"Invalid base metadata for replace transaction, expected non-null"?

Reply:
I was matching them message used directly above, that's why I made the suggestion

Reply:
I included more context above because it isn't clear _what_ was invalid about the metadata. I think the fact that Iceberg expected null is important. Here, the main thing the person reading this needs to know is that metadata was null and that's almost never a good thing.